upload_multipart.go

```diff
@@ -0,0 +1,238 @@
+	cli.mutexUploadID.Lock()
+	defer cli.mutexUploadID.Unlock()
+
+	// List existing Multipart uploads for our bucket
+	listMultiOutput, err := cli.sdkClient.ListMultipartUploads(
+		&s3.ListMultipartUploadsInput{
+			Bucket: &cli.bucketName,
+		})
+	if err != nil {
+		return "", fmt.Errorf("error fetching multipart list: %w", err)
+	}
+
+	// Try to find a multipart upload for the same s3 object that we want
+	for _, upload := range listMultiOutput.Uploads {
+		if *upload.Key == req.UploadKey {
+			uploadID = *upload.UploadId
+			return uploadID, nil
+		}
+	}
+
+	// If we didn't find the Multipart upload, create it
+	createMultiOutput, err := cli.sdkClient.CreateMultipartUpload(
+		&s3.CreateMultipartUploadInput{
+			Bucket:      &cli.bucketName,
+			Key:         &req.UploadKey,
+			ContentType: &req.Type,
+		})
+	if err != nil {
+		return "", fmt.Errorf("error creating multipart upload: %w", err)
+	}
+	return *createMultiOutput.UploadId, nil
+}
+
+// doUploadPart performs the upload using the sdkClient if no psk is provided, or the cryptoClient if psk is provided
+func (cli *Client) doUploadPart(ctx context.Context, input *s3.UploadPartInput, psk []byte) (out *s3.UploadPartOutput, err error) {
+	if psk != nil {
+		// Upload Part with PSK
+		out, err = cli.cryptoClient.UploadPartWithPSK(input, psk)
+		if err != nil {
+			return nil, fmt.Errorf("error uploading part with psk: %w", err)
+		}
+		return
+	}
+
+	// Upload part without user-defined PSK
+	out, err = cli.sdkClient.UploadPart(input)
+	if err != nil {
+		return nil, fmt.Errorf("error uploading part with psk: %w", err)
+	}
+	return
+}
+
+// CheckPartUploaded returns true only if the chunk corresponding to the provided chunkNumber has been uploaded.
+// If all the chunks have been uploaded, we complete the upload operation.
+func (cli *Client) CheckPartUploaded(ctx context.Context, req *UploadPartRequest) (bool, error) {
+	logData := log.Data{
+		"chunk_number": req.ChunkNumber,
+		"max_chunks":   req.TotalChunks,
+		"file_name":    req.FileName,
+		"bucket_name":  cli.bucketName,
+		"identifier":   req.UploadKey,
+	}
+
+	listMultiInput := &s3.ListMultipartUploadsInput{
+		Bucket: &cli.bucketName,
+	}
+
+	listMultiOutput, err := cli.sdkClient.ListMultipartUploads(listMultiInput)
+	if err != nil {
+		return false, NewError(fmt.Errorf("error fetching multipart upload list: %w", err), logData)
+	}
+
+	var uploadID string
+	for _, upload := range listMultiOutput.Uploads {
+		if *upload.Key == req.UploadKey {
+			uploadID = *upload.UploadId
+			break
+		}
+	}
+	if len(uploadID) == 0 {
+		return false, NewError(errors.New("s3 key not uploaded"), logData)
+	}
+
+	input := &s3.ListPartsInput{
+		Key:      &req.UploadKey,
+		Bucket:   &cli.bucketName,
+		UploadId: &uploadID,
+	}
+
+	output, err := cli.sdkClient.ListParts(input)
+	if err != nil {
+		return false, NewError(fmt.Errorf("chunk number verification error: %w", err), logData)
```

Comment:
Is this the correct error string for this context?

Reply:
changed